File: backend/node/src/aws/S3Privado.mjs
import config from "../config.mjs";
import { Upload } from "@aws-sdk/lib-storage";
import fs from "fs";
import {
  S3Client,
  DeleteObjectCommand,
  GetObjectCommand,
} from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";

const client = new S3Client({
  region: config.region_ha,
  credentials: {
    accessKeyId: config.accessKeyId_s3,
    secretAccessKey: config.secretAccessKey_s3,
  },
});

//aquí se sube el archivo a S3, sin utilizar base64
const UploadS3_ = async (filePath, fileName, fileType, folder) => {
  try {
    // Leer el archivo temporal desde el sistema de archivos
    const fileContent = fs.readFileSync(filePath);

    // Subir el archivo a AWS S3
    const upload = new Upload({
      client,
      params: {
        Bucket: config.bucket_s3, //bucket s3
        Key: `${folder}/${fileName}`, //nombre del archivo en s3
        Body: fileContent, //contenido del archivo
        ContentType: fileType, //tipo MIME
      },
    });

    // Borrar el archivo temporal del servidor una vez que se sube a S3
    fs.unlinkSync(filePath);

    const response = await upload.done();
    if (response.$metadata.httpStatusCode !== 200) {
      console.error(
        "Error al subir la imagen. Código de estado:",
        response.$metadata.httpStatusCode
      );
      return false;
    }

    console.log("La imagen se ha subido correctamente.");
    //console.log("URL de la imagen:", response.Location);
    return true;
  } catch (error) {
    console.error("Error al subir el archivo a S3:", error);
    fs.unlinkSync(filePath);
    return false;
  }
};

const generateSignedUrl = async (path) => {
  const getObjectCommand = new GetObjectCommand({
    Bucket: config.bucket_s3,
    Key: path,
  });

  try {
    // Generar URL firmada con tiempo de expiración (ej. 1 hora)
    const signedUrl = await getSignedUrl(client, getObjectCommand, { expiresIn: 3600 });
    return signedUrl;
  } catch (error) {
    console.error("Error al generar la URL firmada:", error);
    return null;
  }
};

const deleteObjectS3 = async (path) => {
  const command = new DeleteObjectCommand({
    Bucket: config.bucket_s3,
    Key: path,
  });

  try {
    const response = await client.send(command);
    return response;
  } catch (error) {
    console.error(error);
    return null;
  }
};

export { UploadS3_, deleteObjectS3, generateSignedUrl };

File: backend/python/src/aws/S3Privado.py
import boto3
import os
from botocore.exceptions import BotoCoreError, ClientError
from config import config

# Crear cliente de S3
s3_client = boto3.client(
    's3',
    region_name=config['region_ha'],
    aws_access_key_id=config['accessKeyId_s3'],
    aws_secret_access_key=config['secretAccessKey_s3']
)

# Subir archivo a S3 desde el sistema de archivos (no base64)
def upload_s3(file_path, file_name, file_type, folder):
    try:
        with open(file_path, 'rb') as file_content:
            s3_client.put_object(
                Bucket=config['bucket_s3'],
                Key=f"{folder}/{file_name}",
                Body=file_content,
                ContentType=file_type
            )

        # Borrar el archivo temporal una vez subido
        os.remove(file_path)

        print("El archivo se ha subido correctamente.")
        return True
    except (BotoCoreError, ClientError) as error:
        print(f"Error al subir el archivo a S3: {error}")
        # Intentar eliminar el archivo temporal si no se subió correctamente
        if os.path.exists(file_path):
            os.remove(file_path)
        return False

# Subir archivo a S3 desde buffer (base64)
def upload_s3_base64(buff, path):
    try:
        s3_client.put_object(
            Bucket=config['bucket_s3'],
            Key=path,
            Body=buff,
            ContentType="image/jpeg"
        )
        return True
    except (BotoCoreError, ClientError) as error:
        print(f"Error al subir el archivo a S3: {error}")
        return False

# Generar una URL firmada con expiración
def generate_signed_url(path, expiration=3600):
    try:
        signed_url = s3_client.generate_presigned_url(
            'get_object',
            Params={'Bucket': config['bucket_s3'], 'Key': path},
            ExpiresIn=expiration
        )
        return signed_url
    except (BotoCoreError, ClientError) as error:
        print(f"Error al generar la URL firmada: {error}")
        return None

# Eliminar objeto de S3
def delete_object_s3(path):
    try:
        response = s3_client.delete_object(
            Bucket=config['bucket_s3'],
            Key=path
        )
        return response
    except (BotoCoreError, ClientError) as error:
        print(f"Error al eliminar el objeto de S3: {error}")
        return None
